### ISSUE TITLE
QA model confidence

### ISSUE THREAD

**Issue by Timoeller** [deepset-ai/FARM#690]
The QA answer score that is currently implemented can be misleading if not downright wrong. Lets create a "score" or confidence that is more informative.

The detailed plan how to approach QA model confidence follows soon.

Related issue on haystack side: https://github.com/deepset-ai/haystack/issues/739

**Comment by julian-risch** [comment]
**First Plan**
We would like a QA model to state something like “with confidence of 75% the predicted answer is correct”.
Here comes a first simple plan based on a technique called temperature scaling. Further improvements could follow later on. The approach works not only for extractive QA but also for classification tasks. An adaptation of the approach might also work for Open Domain QA and RAG.

Current scores given for each generated answer combine the logits of the start and end index into one full answer score but do not really correspond to the model’s confidence. I suggest to scale (“calibrate”) these logits so that the scaled score corresponds to the accuracy (or other metrics) with which start and end indices are predicted correctly on a validation dataset. Thus, the model learns to output confidence scores that are calibrated on the observed confidence on a validation dataset.

The classification tasks (not QA), the idea is also described here: https://arxiv.org/pdf/1706.04599.pdf 
There is an implementation available here: https://github.com/gpleiss/temperature_scaling 

It comes down do adjusting the scores based on the following formula:
`softmax = e^(z/T) / sum_i e^(z_i/T)`
`T` is the temperature for scaling, learned based on the validation dataset
`z` is the logit for start index (or end index) as predicted by the model

For extractive QA, we would end up with two confidence scores, one for the start index and one for the end index. We need to try out different ways of combining them in one single score. 

@lalitpagaria  What do you think about the idea? Would you be interested in collaborating with us on this issue? We could schedule a call to discuss model confidence and the suggested plan. Integration with RAG seems to be more complex. Maybe right after the beam search step, when probabilities for each hypothesis are calculated, is a good point to use temperature scaling and output confidence scores.

**Comment by lalitpagaria** [comment]
Sound good. I need to ramp up these can concept, I will read them and then we will schedule a call.

**Comment by lalitpagaria** [comment]
Hi @julian-risch 
My apologies for late reply. I was held up by some other tasks. Unfortunately I will not get time, as my most of focus now a days on [Obsei](https://github.com/lalitpagaria/obsei). I would love to get your feedback about it.

I have gone through links and topics you shared I am able to understand them (at least Math part).
I see that you already started working on it, I will have better look into your PR to learn more about this topic.

Again I am sorry for ghosting.

**Comment by julian-risch** [comment]
Hi @lalitpagaria no worries. I saw your post on LinkedIn and thought you might be quite busy. If you find time for a call, I'd be happy to walk you through the PR. It is a very first minimal version. Just let me know. In the meantime I will have a closer look at Obsei.

**Comment by lalitpagaria** [comment]
Great. I will send email (mentioned in your Github profile) about call.
Regarding Obsei, I am getting content moderation workflow related enquires and it would be super cool to get you input on it.
